File: packages/python/jobping/trace.py
"""Trace analysis helpers for JobPing diagnostic data.

These are pure functions that consume the ``_trace`` dicts produced when
``wrap_trace()`` is used.  They do not depend on any transport or broker —
you can call them on stored trace payloads offline.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class TraceNode:
    job_id: str
    peer_id: str
    hop: int
    elapsed: float          # total wall-clock including sub-calls
    self_time: float        # own time = elapsed - sum(sub.elapsed)
    sub_jobs: list[TraceNode] = field(default_factory=list)


@dataclass
class TraceReport:
    root: TraceNode
    total_elapsed: float
    critical_path: list[TraceNode]
    bottleneck: TraceNode
    call_graph: dict[str, list[str]]  # job_id → [child_job_ids]


def parse_trace(raw: dict) -> TraceReport:
    """Convert a raw ``_trace`` dict into a structured ``TraceReport``."""
    root = _parse_node(raw)
    return TraceReport(
        root=root,
        total_elapsed=root.elapsed,
        critical_path=_find_critical_path(root),
        bottleneck=_find_bottleneck(root),
        call_graph=_build_adjacency(root),
    )


def find_bottleneck(report: TraceReport) -> str:
    """Return a human-readable bottleneck summary."""
    b = report.bottleneck
    if report.total_elapsed > 0 and b.self_time / report.total_elapsed > 0.5:
        return (
            f"Bottleneck: {b.peer_id} (job {b.job_id}) — "
            f"self_time={b.self_time:.2f}s "
            f"({b.self_time / report.total_elapsed:.0%} of total {report.total_elapsed:.2f}s)"
        )
    return (
        f"Balanced — max self_time is {b.peer_id} "
        f"({b.self_time:.2f}s of {report.total_elapsed:.2f}s total)"
    )


# -- internal helpers ------------------------------------------------------

def _parse_node(raw: dict) -> TraceNode:
    sub_nodes = [_parse_node(sj) for sj in raw.get("sub_jobs", [])]
    elapsed = float(raw.get("elapsed", 0))
    sub_elapsed = sum(sn.elapsed for sn in sub_nodes)
    return TraceNode(
        job_id=raw.get("job_id", "?"),
        peer_id=raw.get("peer_id", "?"),
        hop=int(raw.get("hop", 1)),
        elapsed=elapsed,
        self_time=max(0, elapsed - sub_elapsed),
        sub_jobs=sub_nodes,
    )


def _find_critical_path(node: TraceNode) -> list[TraceNode]:
    """Longest path from *node* to a leaf."""
    path = [node]
    if not node.sub_jobs:
        return path
    longest_child = max(node.sub_jobs, key=lambda n: n.elapsed)
    path.extend(_find_critical_path(longest_child))
    return path


def _find_bottleneck(node: TraceNode) -> TraceNode:
    """Node with the largest self_time in the tree."""
    best = node
    for child in node.sub_jobs:
        candidate = _find_bottleneck(child)
        if candidate.self_time > best.self_time:
            best = candidate
    return best


def _build_adjacency(node: TraceNode) -> dict[str, list[str]]:
    graph: dict[str, list[str]] = {}
    children = [child.job_id for child in node.sub_jobs]
    if children:
        graph[node.job_id] = children
    for child in node.sub_jobs:
        graph.update(_build_adjacency(child))
    return graph

File: packages/js/trace.mjs
// Trace analysis helpers for JobPing diagnostic data.

export class TraceNode {
  constructor({ job_id, peer_id, hop, elapsed, self_time, sub_jobs = [] }) {
    this.job_id = job_id;
    this.peer_id = peer_id;
    this.hop = hop;
    this.elapsed = elapsed;
    this.self_time = self_time;
    this.sub_jobs = sub_jobs;
  }
}

export class TraceReport {
  constructor({ root, total_elapsed, critical_path, bottleneck, call_graph }) {
    this.root = root;
    this.total_elapsed = total_elapsed;
    this.critical_path = critical_path;
    this.bottleneck = bottleneck;
    this.call_graph = call_graph;
  }
}

function parseNode(raw) {
  const subNodes = (raw.sub_jobs || []).map(parseNode);
  const elapsed = parseFloat(raw.elapsed) || 0;
  const subElapsed = subNodes.reduce((sum, sn) => sum + sn.elapsed, 0);
  return new TraceNode({
    job_id: raw.job_id || "?",
    peer_id: raw.peer_id || "?",
    hop: parseInt(raw.hop, 10) || 1,
    elapsed,
    self_time: Math.max(0, elapsed - subElapsed),
    sub_jobs: subNodes,
  });
}

function findCriticalPath(node) {
  const path = [node];
  if (!node.sub_jobs.length) return path;
  const longestChild = node.sub_jobs.reduce((a, b) => (a.elapsed > b.elapsed ? a : b));
  path.push(...findCriticalPath(longestChild));
  return path;
}

function findBottleneckNode(node) {
  let best = node;
  for (const child of node.sub_jobs) {
    const candidate = findBottleneckNode(child);
    if (candidate.self_time > best.self_time) best = candidate;
  }
  return best;
}

function buildAdjacency(node) {
  const graph = {};
  if (node.sub_jobs.length) {
    graph[node.job_id] = node.sub_jobs.map((c) => c.job_id);
  }
  for (const child of node.sub_jobs) {
    Object.assign(graph, buildAdjacency(child));
  }
  return graph;
}

export function parseTrace(raw) {
  const root = parseNode(raw);
  return new TraceReport({
    root,
    total_elapsed: root.elapsed,
    critical_path: findCriticalPath(root),
    bottleneck: findBottleneckNode(root),
    call_graph: buildAdjacency(root),
  });
}

export function findBottleneck(report) {
  const b = report.bottleneck;
  if (report.total_elapsed > 0 && b.self_time / report.total_elapsed > 0.5) {
    return (
      `Bottleneck: ${b.peer_id} (job ${b.job_id}) — ` +
      `self_time=${b.self_time.toFixed(2)}s ` +
      `(${(b.self_time / report.total_elapsed * 100).toFixed(0)}% of total ${report.total_elapsed.toFixed(2)}s)`
    );
  }
  return (
    `Balanced — max self_time is ${b.peer_id} ` +
    `(${b.self_time.toFixed(2)}s of ${report.total_elapsed.toFixed(2)}s total)`
  );
}
